utils/zerovec/src/zerovec/ule.rs

```diff
@@ -0,0 +1,160 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use alloc::boxed::Box;
+
+/// A type similar to [`ZeroVec`] that can be used inside `VarZeroVec<T>`.
+///
+/// This essentially allows for the construction of zero-copy types isomorphic to
+/// `Vec<Vec<T>>` by instead using `VarZeroVec<ZeroVecULE<T>>`.
+///
+/// See the [`VarZeroVec`] docs for an example.
+#[repr(transparent)]
+pub struct ZeroVecULE<T: AsULE>(pub [T::ULE]);
+
+impl<T> ZeroVecULE<T>
+where
+    T: AsULE,
+{
+    /// Get this [`ZeroVecULE`] as a borrowed [`ZeroVec`]
+    ///
+    /// [`ZeroVecULE`] does not have most of the methods that [`ZeroVec`] does,
+    /// so it is recommended to convert it to a [`ZeroVec`] before doing anything.
+    #[inline]
+    pub fn as_zerovec(&self) -> ZeroVec<'_, T> {
+        ZeroVec::Borrowed(&self.0)
```

Comment:
It occurs to me that `&ZVULE` is a perfectly acceptable ZVBorrowed :smile:

(and perhaps VZVBorrowed is unnecessary too; we could rename them to ZeroSlice and VarZeroSlice, even)

Reply:
actually I _really_ like the idea of ZeroSlice and VarZeroSlice because it's great symmetry

Reply:
(I won't do that here, though, perhaps as a followup)

Reply:
Yeah, I like that idea!  To be clear, `ZeroSlice` would be what you currently have as `ZeroVecULE`, correct?  So then it would always be a `&ZeroSlice` or `Box<ZeroSlice>`?

Even better: make `ZeroVec` deref to `ZeroSlice` and VZV to VZS, and then we can finally unify the impls of the methods, I think!

Reply:
Yes and yes!


Slight hitch with dereffing on VZV is that currently things work off of VZB internally and VZB is partially pre-parsed for efficiency.

Reply:
To help a little bit, you could make it

```
struct VarZeroSlice {
    length: u32::ULE,
    everything_else: [u8]
}
```

Reply:
You could maybe even consider

```
enum VarZeroVec<'a> {
    Borrowed(&'a VarZeroSlice),
    Owned(Box<VarZeroSlice>)
}
```

The mutation methods would need to do some gymnastics, but you can't deny that the signature is a lot cleaner

Reply:
_hmmm_ yes that works, still wary of introducing a bunch of conversions to the common case

Reply:
I'm filing an issue for this, but I suspect what I want to do here is:

 - Perform the rename for both, use `&ZVS` instead of `&[T::ULE]` in a bunch of places
 - Add deref impls but do _not_ deduplicate most of the methods on VZV, let the inherent methods be a transparent perf boost
 - Measure performance and unify VZVS/VZVB if necessary.

Reply:
The "conversion" is only an unaligned read of the length relative to what you have now.  We can at least avoid doing the bounds check.  I think that's a fairly acceptable trade-off, although Criterion can tell us for sure.

Reply:
No, we also construct the indices array, the LLVM ir is nontrivial

Reply:
There's a bit more going on in the IR I see but yeah we can measure.

Filed https://github.com/unicode-org/icu4x/issues/1367 so we can continue the discussion there. I can at least proceed with the renames and cleanups.